File: app/src/main/java/com/example/domain/repo/AuthRepository.kt
package com.example.domain.repo

import com.example.domain.model.AuthState
import com.example.domain.model.AuthChallenge
import kotlinx.coroutines.flow.Flow

/**
 * Interface safeguarding X OAuth 2.0 PKCE authentication lifecycle and session credentials.
 */
interface AuthRepository {
    /**
     * Observable stream representing current authentication credentials and session status.
     */
    fun authState(): Flow<AuthState>

    /**
     * Initializes a PKCE challenge state, returning details required for launching the login sheet.
     */
    suspend fun beginLogin(): AuthChallenge

    /**
     * Exchanges auth code using original code verifier from previous flow.
     */
    suspend fun completeLogin(code: String, codeVerifier: String): Result<Unit>

    /**
     * Resolves currently active user's numeric ID (for retrieving bookmarks safely).
     */
    suspend fun currentUserId(): String?

    /**
     * Purges auth credentials silently, notifying all sessions.
     */
    suspend fun logout()

    /**
     * Best-effort backfill of public profile fields (handle, display name, profile photo) for a
     * session that predates their storage — e.g. users who signed in before the profile photo was
     * fetched. No-op when signed out or when the photo is already cached; never throws.
     */
    suspend fun refreshProfile()
}

File: ios/Curio/Curio/Domain/AuthRepository.swift
import Foundation
import Combine

/// Interface safeguarding X OAuth 2.0 PKCE authentication lifecycle and session credentials. Ports
/// `interface AuthRepository` from `domain/repo/AuthRepository.kt`.
///
/// Concurrency mapping (CONVENTIONS §3 / §11):
/// - Kotlin hot `Flow<AuthState>` → Combine `AnyPublisher<AuthState, Never>` (re-emits on every
///   session change; backed by a `CurrentValueSubject` in the impl).
/// - `suspend fun beginLogin(): AuthChallenge` → `async throws` (it can fail while building the
///   challenge / generating PKCE).
/// - `suspend fun completeLogin(...): Result<Unit>` → `async throws` (the one auth path that
///   surfaces a typed failure to the UI — see CONVENTIONS §"Resilience contract").
/// - plain `suspend fun currentUserId(): String?` → `async` (NO throws).
/// - plain `suspend fun logout()` → `async` (NO throws).
///
/// `Sendable` because the repository is used across actor boundaries.
protocol AuthRepository: Sendable {
    /// Observable stream representing current authentication credentials and session status.
    func authState() -> AnyPublisher<AuthState, Never>

    /// Initializes a PKCE challenge, returning details required for launching the login sheet.
    func beginLogin() async throws -> AuthChallenge

    /// Exchanges the auth code using the original code verifier from the previous flow.
    /// Throws on failure (collapses Kotlin `Result<Unit>`).
    func completeLogin(code: String, codeVerifier: String) async throws

    /// Resolves the currently active user's numeric ID (for retrieving bookmarks safely).
    func currentUserId() async -> String?

    /// Purges auth credentials silently, notifying all sessions.
    func logout() async
}
